adam/learner/propose_but_verify.py

```diff
@@ -0,0 +1,225 @@
+        Create a new `Concept` of the appropriate type with the given *debug_string*.
+        """
+
+    @abstractmethod
+    def _hypotheses_from_perception(
+        self,
+        learning_state: LanguagePerceptionSemanticAlignment,
+        bound_surface_template: SurfaceTemplateBoundToSemanticNodes,
+    ) -> AbstractSet[PerceptionGraphTemplate]:
+        """
+        Get a hypothesis for the meaning of *surface_template* from a given *learning_state*.
+        """
+
+    def _primary_templates(
+        self
+    ) -> Iterable[Tuple[Concept, PerceptionGraphTemplate, float]]:
+        return (
+            (concept, hypotheses[0], 1.0)
+            for (concept, hypotheses) in self._concept_to_hypotheses.items()
+            # We are confident in a hypothesis if we don't have any alternatives.
+            if len(hypotheses) == 1
+        )
+
+    def _fallback_templates(
+        self
+    ) -> Iterable[Tuple[Concept, PerceptionGraphTemplate, float]]:
+        # Alternate hypotheses stored in the beam.
+        return (
+            (concept, hypothesis, 1.0)
+            for (concept, hypotheses) in self._concept_to_hypotheses.items()
+            for hypothesis in hypotheses[1:]
+            if len(hypotheses) > 1
+        )
+
+    def _match_template(
+        self,
+        *,
+        concept: Concept,
+        pattern: PerceptionGraphTemplate,
+        perception_graph: PerceptionGraph,
+    ) -> Optional[Tuple[PerceptionGraphPatternMatch, SemanticNode]]:
+        """
+        Try to match our model of the semantics to the perception graph
+        """
+        partial_match = compute_match_ratio(
+            pattern,
+            perception_graph,
+            ontology=self._ontology,
+            graph_logger=self._graph_logger,
+            debug_callback=self._debug_callback,
+        )
+
+        if (
+            partial_match.match_ratio >= self._minimum_match_ratio
+            and partial_match.matching_subgraph
+        ):
+            # if there is a match, which is above our minimum match ratio
+            # Use that pattern to try and find a match in the scene
+            # There should be one
+            # TODO: This currently means we match to the graph multiple times. Reduce this?
+            matcher = partial_match.matching_subgraph.matcher(
+                perception_graph,
+                match_mode=MatchMode.NON_OBJECT,
+                debug_callback=self._debug_callback,
+            )
+            for match in matcher.matches(use_lookahead_pruning=True):
+                semantic_node_for_match = pattern_match_to_semantic_node(
+                    concept=concept, pattern=pattern, match=match
+                )
+                # A template only has to match once; we don't care about finding additional matches.
+                return match, semantic_node_for_match
+            # We raise an error if we find a partial match but don't manage to match it to the scene
+            raise RuntimeError(
```

Comment:
To we want to raise an error here or just refuse to learn?

Reply:
I'd prefer to raise an error here because this case would mean we identified a partial match in the scene which matches our hypothesis enough to say "Yes this is the object we think it is" but we were unable to match the reduced pattern to the perception -- Which would be an error.